sda/cmd/rotatekey/rotatekey.go

```diff
@@ -0,0 +1,268 @@
+		panic(err)
+	}
+	db, err = database.NewSDAdb(conf.Database)
+	if err != nil {
+		panic(err)
+	}
+
+	go func() {
+		<-sigc // blocks here until it receives from sigc
+		fmt.Println("Interrupt signal received. Shutting down.")
+		defer mq.Channel.Close()
+		defer mq.Connection.Close()
+		defer db.Close()
+
+		os.Exit(0) // exit program
+	}()
+
+	// encode pubkey as pem and then as base64 string
+	tmp := &bytes.Buffer{}
+	if err := keys.WriteCrypt4GHX25519PublicKey(tmp, *conf.RotateKey.PublicKey); err != nil {
+		panic(err)
+	}
+	pubKeyEncoded := base64.StdEncoding.EncodeToString(tmp.Bytes())
+
+	// Check that key is registered in the db at startup
+	err = db.CheckKeyHash(hex.EncodeToString(conf.RotateKey.PublicKey[:]))
+	if err != nil {
+		panic(fmt.Errorf("database lookup of the rotation key failed, reason: %v", err))
+	}
+
+	go func() {
+		connError := mq.ConnectionWatcher()
+		log.Error(connError)
+		forever <- false
+	}()
+
+	go func() {
+		connError := mq.ChannelWatcher()
+		log.Error(connError)
+		forever <- false
+	}()
+
+	log.Info("Starting rotatekey service")
+	var message schema.KeyRotation
+
+	go func() {
+		// Create a function to handle panic and exit gracefully
+		defer func() {
+			if err := recover(); err != nil {
+				if mq != nil {
+					defer mq.Channel.Close()
+					defer mq.Connection.Close()
+				}
+				if db != nil {
+					defer db.Close()
+				}
+				log.Fatal(err)
+			}
+		}()
+		messages, err := mq.GetMessages(conf.Broker.Queue)
+		if err != nil {
+			panic(err)
+		}
+		for delivered := range messages {
+			log.Debugf("Received a message (corr-id: %s, message: %s)",
+				delivered.CorrelationId,
+				delivered.Body)
+
+			err := schema.ValidateJSON(fmt.Sprintf("%s/rotate-key.json", conf.Broker.SchemasPath), delivered.Body)
+			if err != nil {
+				msg := "validation of incoming message (rotate-key) failed"
+				log.Errorf("%s, reason: %v", msg, err)
+				nackAndSendToErrorQueueckAndSendToErrorQueue(mq, delivered, conf.Broker.Exchange, msg, err.Error())
+
+				continue
+			}
+
+			// Fetch rotate key hash before starting work so that we make sure the hash state
+			// has not changed since the application startup.
+			keyhash := hex.EncodeToString(conf.RotateKey.PublicKey[:])
+			// exit app if target key was modified after app start-up, e.g. if key has been deprecated
+			if err = db.CheckKeyHash(keyhash); err != nil {
+				panic(fmt.Errorf("check of target key failed, reason: %v", err))
+			}
+
+			// we unmarshal the message in the validation step so this is safe to do
+			_ = json.Unmarshal(delivered.Body, &message)
+
+			fileID := message.FileID
+
+			// Get current keyhash for the file, send to error queue if this fails
+			oldKeyHash, err := db.GetKeyHash(fileID)
+			if err != nil {
+				msg := fmt.Sprintf("failed to get keyhash for file with file-id: %s", fileID)
+				log.Errorf("%s, reason: %v", msg, err)
+				nackAndSendToErrorQueueckAndSendToErrorQueue(mq, delivered, conf.Broker.Exchange, msg, err.Error())
```

Comment:
This should have separate handlings dependent on the error.
type 1 errors: nack and requeue the message so we can retry later.
type 2 errors: ack and send to error queue

Reply:
...which would both lead to empty rows.